flybywiresim/a32nx
# feat(isis): ISIS auto brightness

Fixes #5101

## Summary of Changes
Automatically set ISIS (Integrated Standby Instrument System) to appropriate brightness based on time of day. The real aircraft does this using an ambient light sensor. To my knowledge, there is no way to extract this value in the simulator though, so we do what a simulator does and simulate it by checking whether it is daytime where the airplane is which typically means that sunlight falls into the cockpit causing the ISIS to brighten up. We check for daytime using the `E:TIME OF DAY` SimVar. This is the same logic the automatic brightness control the chrono instrument uses (#5018). When a change in the time of day, the brightness setting transitions from to opposite setting.

No change in brightness will be observed if ambient light changes in the cockpit by other means that time of day (dome lights, clouds, etc.)

The manual brightness buttons still work by adjusting the increasing/decreasing the current brightness by some granularity. The manually set brightness is then taken over as the new day/night preset value. If the brightness is adjusted during a transition, the same happens and the transition is halted.

It is quite simple for me to adjust levels of brightness for the default day/night presets if desired.

## Screenshots (if necessary)


Before:
![isis-before](https://user-images.githubusercontent.com/22713769/122655905-86916400-d156-11eb-87c4-11987f1fd132.png)
After:
![isis-after](https://user-images.githubusercontent.com/22713769/122655910-8b561800-d156-11eb-9a12-a799f41ca296.png)

## References

- AntoineXWB who opened issue 5101 is an Airbus engineer and provided information on how the system works in the real aircraft. We try to simulate this.




## Additional context



Discord username (if different from GitHub): BlueberryKing#6641

## Testing instructions

- Spawn in (C/D or on the runway). 
- The ISIS should have a reasonable brightness level.
- Use the in-game menu to change the time of day either from day to night or night to day.
- Observe how the brightness transitions to the new brightness setting (the transition should take about 7 seconds, but it can take up to 10 seconds to start). Again, the ISIS should be at a reasonable setting for this time of day.
- Manually change the brightness of the instrument using the buttons to the left. This should behave as it previously has; adjusting brightness between a minimum and maximum. Holding down a button should also behave as if the button were repeatedly pressed.
- For maximum effect, set an extreme brightness setting manually (either full brightness or full darkness).
- Change time of day and observe that instrument is in extreme setting it had previously been set to manually.


## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page
